### PR TITLE
ECO:340- adding GetAlertNotifications

### PR DESCRIPTION
**Pull request by rshubert**

### CODE REVIEW

File: Apps/phironnet/ironnet.json [hunk 1]
```diff
@@ -41,6 +41,48 @@
...
             "required": true,
             "order": 3
+        },
+        "enable_alert_notifications": {
+            "description": "Enable Alert Notifications",
+            "data_type": "boolean",
+            "required": false,
+            "default": true,
+            "order": 10
+        },
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json:58]: Add a configuration for the alert_action(s) to consume - defaulting to just ALERT_CREATED.

**Review by mjsell12 — CHANGES_REQUESTED** [review]

File: Apps/phironnet/ironnet_connector.py [hunk 2]
```diff
@@ -319,6 +325,51 @@ def _handle_irondefense_report_observed_bad_activity(self, param):
...
+            # make rest call
+            ret_val, response = self._make_post('/GetAlertNotifications', action_result, data=request, headers=None)
+            if phantom.is_success(ret_val):
+                self.save_progress("Fetching alert notifications was successful")
+                # Filter the response and add into the data section
+                for alert_notification in response['alert_notifications']:
+                    if alert_notification['alert']:
+                        alert = alert_notification['alert']
+                        if alert['category'] not in alert_cats and alert['sub_category'] not in alert_subcats:
+                            if alert_sev_lower <= int(alert['severity']) <= alert_sev_upper:
```
Comment (marcus-defazio) [inline Apps/phironnet/ironnet_connector.py]: Could you roll this if statement into the previous one? There is no need to nest this statement.

File: Apps/phironnet/ironnet_connector.py [hunk 3]
```diff
@@ -319,6 +325,51 @@ def _handle_irondefense_report_observed_bad_activity(self, param):
...
+                                contnr = {
+                                    'name': str(alert["id"]) + ' Alert Notification',
```
Comment (marcus-defazio) [inline Apps/phironnet/ironnet_connector.py]: This should be be changed to
```suggestion
                                    'name': '{} Alert Notification'.format(alert_notification['alert_action']),
```
so it is consistent with Demisto.

File: Apps/phironnet/ironnet_connector.py [hunk 4]
```diff
@@ -319,6 +325,51 @@ def _handle_irondefense_report_observed_bad_activity(self, param):
...
+                                    'description': "Ingested {} Alert Notification occurring at {}.".
```
Comment (marcus-defazio) [inline Apps/phironnet/ironnet_connector.py]: This wording should also be changed to match Demisto.

File: Apps/phironnet/ironnet_connector.py [hunk 5]
```diff
@@ -319,6 +325,51 @@ def _handle_irondefense_report_observed_bad_activity(self, param):
...
+                                    format(alert_notification["alert_action"], str(alert["updated"])),
+                                    'source_data_identifier': str(alert['id']) + '_' + str(alert['event_count']),
```
Comment (marcus-defazio) [inline Apps/phironnet/ironnet_connector.py]: What is the `source_data_identifier` and why are we appending the event count to the alert ID?

File: Apps/phironnet/ironnet_connector.py [hunk 6]
```diff
@@ -353,6 +409,26 @@ def initialize(self):
         self._password = config.get('password')
         self._verify_server_cert = config.get('verify_server_cert')
 
+        # Alert Notification Configs
+        self._enable_alert_notifications = config.get('enable_alert_notifications')
+        alert_cats = config.get('alert_categories')
+        if alert_cats:
+            self._alert_categories = [str(cat).replace(" ", "_").upper() for cat in alert_cats.split(',')]
```
Comment (marcus-defazio) [inline Apps/phironnet/ironnet_connector.py]: We should also trim whitespace off each parsed alert category to minimize user error.

File: Apps/phironnet/ironnet_connector.py [hunk 7]
```diff
@@ -353,6 +409,26 @@ def initialize(self):
...
+        else:
+            self._alert_categories = []
+        alert_subcats = config.get('alert_subcategories')
+        if alert_subcats:
+            self._alert_subcategories = [str(subcat).replace(" ", "_").upper() for subcat in alert_subcats.split(',')]
```
Comment (marcus-defazio) [inline Apps/phironnet/ironnet_connector.py]: We should also trim whitespace off each parsed alert subcategory to minimize user error.

**Review by marcus-defazio — CHANGES_REQUESTED** [review]

File: Apps/phironnet/ironnet.json [hunk 8]
```diff
@@ -41,6 +41,55 @@
             "data_type": "password",
             "required": true,
             "order": 3
+        },
+        "enable_alert_notifications": {
+            "description": "Enable Alert Notifications",
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json]:
```suggestion
            "description": "Ingest IronDefense Alert Notifications into Phantom",
```

File: Apps/phironnet/ironnet.json [hunk 9]
```diff
@@ -41,6 +41,55 @@
...
+            "data_type": "boolean",
+            "required": false,
+            "default": true,
+            "order": 4
+        },
+        "alert_notif_actions": {
+            "description": "Alert Notification Actions to Include (enter in CSV format)",
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json]:
```suggestion
            "description": "Alert Notification Actions to include for ingest (if ingest is enabled).  Enter in CSV format.",
```

File: Apps/phironnet/ironnet.json [hunk 10]
```diff
@@ -41,6 +41,55 @@
...
+            "data_type": "string",
+            "required": false,
+            "default": "Alert Created",
+            "order": 5
+        },
+        "alert_categories": {
+            "description": "Alert Categories to Exclude (enter in CSV format)",
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json]:
```suggestion
            "description": "Alert Categories to exclude from ingest (if ingest is enabled).  Enter in CSV format.",
```

File: Apps/phironnet/ironnet.json [hunk 11]
```diff
@@ -41,6 +41,55 @@
...
+            "data_type": "string",
+            "required": false,
+            "default": "",
+            "order": 66
+        },
+        "alert_subcategories": {
+            "description": "Alert SubCategories to Exclude (enter in CSV format)",
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json]:
```suggestion
            "description": "Alert SubCategories to exclude from ingest (if ingest is enabled). Enter in CSV format.",
```

File: Apps/phironnet/ironnet_connector.py (hunk 5 above)
Comment (rshubert) [reply Apps/phironnet/ironnet_connector.py]: source_data_identifier is a the id from the source (so alert id in this case), but it has to be unique which is why i tried adding the event count. but changed how we're storing the data now

File: Apps/phironnet/ironnet.json [hunk 12]
```diff
@@ -41,6 +41,55 @@
...
+            "data_type": "string",
+            "required": false,
+            "default": "",
+            "order": 7
+        },
+        "alert_severity_lower": {
+            "description": "Minimum Severity of Alerts",
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json]:
```suggestion
            "description": "Minimum Severity of Alerts to ingest (if ingest is enabled)",
```

File: Apps/phironnet/ironnet.json [hunk 13]
```diff
@@ -41,6 +41,55 @@
...
+            "data_type": "numeric",
+            "required": false,
+            "default": 500,
+            "order": 8
+        },
+        "alert_severity_upper": {
+            "description": "Maximum Severity of Alerts",
```
Comment (mjsell12) [inline Apps/phironnet/ironnet.json]:
```suggestion
            "description": "Maximum Severity of Alerts to ingest (if ingest is enabled)",
```

File: Apps/phironnet/ironnet_connector.py (hunk 4 above)
Comment (rshubert) [reply Apps/phironnet/ironnet_connector.py]: recent refactor changes this, containers are now just the alerts and the notifications are the artifacts of the container

File: Apps/phironnet/ironnet_connector.py [hunk 14]
```diff
@@ -345,6 +352,67 @@ def _handle_irondefense_get_alert_irondome_info(self, param):
...
+            if phantom.is_success(ret_val):
+                self.save_progress("Fetching alert notifications was successful")
+                # Filter the response and add into the data section
+                for alert_notification in response['alert_notifications']:
+                    if alert_notification['alert_action'] in self._alert_notif_actions and alert_notification['alert']:
+                        alert = alert_notification['alert']
+                        if alert['category'] not in self._alert_categories and alert['sub_category'] not in self._alert_subcategories:
+                            if self._alert_severity_lower <= int(alert['severity']) <= self._alert_severity_upper:
+                                # create container
+                                contnr = {
```
Comment (mjsell12) [inline Apps/phironnet/ironnet_connector.py]: Not sure why - but `contnr` really bugs me.  Can we rename to `container` please? :)

**Review by mjsell12 — CHANGES_REQUESTED** [review]

File: Apps/phironnet/ironnet_connector.py (hunk 2 above)
Comment (rshubert) [reply Apps/phironnet/ironnet_connector.py]: the compiler fails because the line is too long

**Review by mjsell12 — APPROVED** [review]

**Pull request merged**